File: crates/chezmoi-support/src/command.rs
use std::path::Path;

use crate::error::{Error, Result};
use crate::fs::write_text_if_changed;

pub fn write_command_text_if_available(
    bin: &str,
    path: &Path,
    command: &duct::Expression,
) -> Result<bool> {
    if which::which(bin).is_err() {
        return Ok(false);
    }
    let text = command_text(command)?;
    write_text_if_changed(path, &text)
}

pub fn command_text(command: &duct::Expression) -> Result<String> {
    let output = command
        .stdout_capture()
        .stderr_capture()
        .unchecked()
        .run()?;
    if !output.status.success() {
        return Err(Error::CommandFailed(format!("{command:?}")));
    }
    Ok(String::from_utf8_lossy(&output.stdout).into_owned())
}

pub fn run_command(command: &duct::Expression) -> Result<()> {
    let output = command.unchecked().run()?;
    if output.status.success() {
        Ok(())
    } else {
        Err(Error::CommandFailed(format!("{command:?}")))
    }
}

pub fn warn_if_failed(name: &str, output: &std::process::Output) {
    if output.status.success() {
        return;
    }
    let message = String::from_utf8_lossy(&output.stderr);
    let message = message.trim();
    if message.is_empty() {
        eprintln!("warn: failed to generate {name} completions");
    } else {
        eprintln!("warn: failed to generate {name} completions: {message}");
    }
}

File: internal/chezmoi/command.go
package chezmoi

import (
	"fmt"
	"strings"

	"github.com/charmbracelet/log"
	"github.com/euvlok/nix-dotfiles/internal/common/fileutil"
	"github.com/euvlok/nix-dotfiles/internal/common/process"
)

func writeCommandTextIfAvailable(bin, path string, argv []string) error {
	if _, ok := process.PathOf(bin); !ok {
		return nil
	}
	text, err := commandText(argv)
	if err != nil {
		return err
	}
	_, err = fileutil.WriteTextIfChanged(path, text)
	return err
}

func commandText(argv []string) (string, error) {
	output, err := process.CaptureWithEnvAndStdin(argv, nil, nil)
	if err != nil {
		return "", err
	}
	if !output.Success {
		label := "<empty>"
		if len(argv) > 0 {
			label = argv[0]
		}
		return "", fmt.Errorf("command failed: %s", label)
	}
	return string(output.Stdout), nil
}

func warnIfFailed(name string, output process.Output) {
	if output.Success {
		return
	}
	message := strings.TrimSpace(string(output.Stderr))
	if message == "" {
		message = strings.TrimSpace(string(output.Stdout))
	}
	if message == "" {
		log.Warn("failed to generate completions", "name", name)
	} else {
		log.Warn("failed to generate completions", "name", name, "message", message)
	}
}
